File: apps/web/src/lib/training/readiness.ts
/**
 * Readiness-to-run score. Combines training balance (form / TSB),
 * last night's sleep, and resting-HR drift vs a baseline into a
 * single 0–100 number the dashboard can show.
 *
 * Garmin's "Body Battery", Whoop's "Recovery" and Oura's "Readiness"
 * are the precedents. We compute deltas around a neutral baseline of
 * 75; positive contributors push it up, negative ones drag it down,
 * and the result is clamped to 0–100.
 *
 * Pure — no Supabase, no DOM. TS↔Dart parity pair with
 * `apps/mobile_android/lib/readiness.dart`. Keep in lockstep — the
 * shared-library-syncer agent watches the pair.
 */

export interface ReadinessInputs {
	/** Training Stress Balance (Form). Positive = fresh, negative = fatigued. Pass null if unknown. */
	tsb: number | null;
	/** Hours of sleep last night. Pass null if unknown. */
	sleepHours?: number | null;
	/** This morning's resting heart rate, in bpm. Pass null if unknown. */
	restingHrBpm?: number | null;
	/** 30-day baseline resting heart rate. Pass null if unknown. */
	baselineRestingHrBpm?: number | null;
}

export type ReadinessBand = 'low' | 'moderate' | 'high';

/**
 * Which input a contribution came from. The UI names it in the reader's
 * locale. The form input is deliberately not labelled like the TSB tile it is
 * derived from: this is the POINTS form added to the score, not the TSB value,
 * and one name over two different numbers read as a contradiction (#902).
 */
export type ReadinessContributorKind = 'form' | 'sleep' | 'resting_hr';

export interface ReadinessContribution {
	kind: ReadinessContributorKind;
	/** Signed delta this input added to the score. */
	delta: number;
	/** One-line reason. */
	note: string;
}

export interface Readiness {
	/** Final score in 0..100. */
	score: number;
	/** Bucket — drives card colour and tone of advice. */
	band: ReadinessBand;
	/** One-line guidance, picked from the dominant contributor. */
	advice: string;
	/** Per-input breakdown so the UI can show "why". */
	contributors: ReadinessContribution[];
}

/** Neutral starting point. A user with no inputs at all gets 75. */
const BASELINE_SCORE = 75;

function clamp(v: number, lo: number, hi: number): number {
	return Math.min(hi, Math.max(lo, v));
}

function bandFor(score: number): ReadinessBand {
	if (score >= 70) return 'high';
	if (score >= 40) return 'moderate';
	return 'low';
}

/**
 * Score the contribution from Training Stress Balance.
 *
 * TSB convention (matches `lib/training/training_load.ts`):
 *  - TSB ≈ 0   → neutral / maintenance form
 *  - TSB > +5  → fresh; small positive on score
 *  - TSB < -10 → fatigued; large negative on score
 *  - TSB > +25 → over-tapered, slightly negative again ("detraining
 *               risk" / blunted edge — small but real)
 */
function scoreTsb(tsb: number | null): ReadinessContribution | null {
	if (tsb == null) return null;
	let delta = 0;
	let note = '';
	if (tsb < -20) {
		delta = -20;
		note = 'Heavy fatigue — recent training stress is well above your fitness';
	} else if (tsb < -10) {
		delta = -12;
		note = 'Fatigued from recent training';
	} else if (tsb < -5) {
		delta = -6;
		note = 'Slight fatigue';
	} else if (tsb <= 5) {
		delta = 0;
		note = 'Form is neutral';
	} else if (tsb <= 15) {
		delta = 8;
		note = 'Fresh and well-recovered';
	} else if (tsb <= 25) {
		delta = 5;
		note = 'Very fresh — borderline tapered';
	} else {
		delta = -3;
		note = 'Over-tapered — edge may be blunted';
	}
	return { kind: 'form', delta, note };
}

function scoreSleep(hours: number | null | undefined): ReadinessContribution | null {
	if (hours == null) return null;
	let delta = 0;
	let note = '';
	if (hours < 5) {
		delta = -25;
		note = 'Very little sleep — recovery is compromised';
	} else if (hours < 6.5) {
		delta = -12;
		note = 'Short on sleep';
	} else if (hours < 7.5) {
		delta = -3;
		note = 'A little under target sleep';
	} else if (hours <= 9) {
		delta = 5;
		note = 'Well rested';
	} else {
		delta = 0;
		note = 'Extended sleep — recovery should be solid';
	}
	return { kind: 'sleep', delta, note };
}

function scoreRestingHr(
	resting: number | null | undefined,
	baseline: number | null | undefined,
): ReadinessContribution | null {
	if (resting == null || baseline == null) return null;
	const diff = resting - baseline;
	let delta = 0;
	let note = '';
	if (diff > 10) {
		delta = -18;
		note = `Resting HR ${diff} bpm above baseline — strong sign of illness or under-recovery`;
	} else if (diff > 5) {
		delta = -10;
		note = `Resting HR ${diff} bpm above baseline`;
	} else if (diff > 2) {
		delta = -4;
		note = 'Resting HR slightly elevated';
	} else if (diff >= -2) {
		delta = 0;
		note = 'Resting HR at baseline';
	} else {
		delta = 3;
		note = 'Resting HR below baseline — well recovered';
	}
	return { kind: 'resting_hr', delta, note };
}

/** Pick the contributor that pushed the score the most (in absolute terms). */
function dominantAdvice(contributors: ReadinessContribution[], band: ReadinessBand): string {
	if (contributors.length === 0) {
		// No signals at all. Fall back to a band-aware suggestion.
		return band === 'high'
			? 'Looks like a good day to push the pace.'
			: 'Connect sleep + HR data for a real readiness picture.';
	}
	// Ties in |delta| are common — a TSB of -12 and six hours of sleep both
	// score -12, and the two carry DIFFERENT advice — so which one wins is a
	// contract, not an accident. `Array.prototype.sort` is stable, so the
	// first contributor added wins; the Dart twin's `List.sort` is NOT, and
	// carries an explicit index tiebreak to match.
	const dom = contributors.slice().sort((a, b) => Math.abs(b.delta) - Math.abs(a.delta))[0];
	const tail =
		band === 'low'
			? 'Consider an easy day or a rest day.'
			: band === 'moderate'
				? 'A steady or moderate effort fits today.'
				: 'You’re primed for a harder effort if planned.';
	return `${dom.note}. ${tail}`;
}

export function computeReadiness(inputs: ReadinessInputs): Readiness {
	const contributors: ReadinessContribution[] = [];
	const tsb = scoreTsb(inputs.tsb);
	if (tsb) contributors.push(tsb);
	const sleep = scoreSleep(inputs.sleepHours ?? null);
	if (sleep) contributors.push(sleep);
	const hr = scoreRestingHr(
		inputs.restingHrBpm ?? null,
		inputs.baselineRestingHrBpm ?? null,
	);
	if (hr) contributors.push(hr);

	const sum = contributors.reduce((s, c) => s + c.delta, 0);
	const score = clamp(Math.round(BASELINE_SCORE + sum), 0, 100);
	const band = bandFor(score);
	return {
		score,
		band,
		advice: dominantAdvice(contributors, band),
		contributors,
	};
}

File: apps/mobile_android/lib/readiness.dart
/// Readiness-to-run score. Combines training balance (form / TSB),
/// last night's sleep, and resting-HR drift vs a baseline into a
/// single 0–100 number the dashboard can show.
///
/// TS↔Dart parity pair with `apps/web/src/lib/training/readiness.ts`. Keep in
/// lockstep — the shared-library-syncer agent watches the pair.

class ReadinessInputs {
  /// Training Stress Balance (Form). Positive = fresh, negative = fatigued.
  final double? tsb;
  final double? sleepHours;
  final int? restingHrBpm;
  final int? baselineRestingHrBpm;

  const ReadinessInputs({
    this.tsb,
    this.sleepHours,
    this.restingHrBpm,
    this.baselineRestingHrBpm,
  });
}

enum ReadinessBand { low, moderate, high }

/// Which input a contribution came from. The UI names it in the reader's
/// locale. The form input is deliberately not labelled like the TSB stat it is
/// derived from: this is the POINTS form added to the score, not the TSB value,
/// and one name over two different numbers read as a contradiction (#902).
enum ReadinessContributorKind { form, sleep, restingHr }

class ReadinessContribution {
  final ReadinessContributorKind kind;
  final int delta;
  final String note;
  const ReadinessContribution({
    required this.kind,
    required this.delta,
    required this.note,
  });
}

class Readiness {
  final int score;
  final ReadinessBand band;
  final String advice;
  final List<ReadinessContribution> contributors;
  const Readiness({
    required this.score,
    required this.band,
    required this.advice,
    required this.contributors,
  });
}

const int _baselineScore = 75;

int _clamp(int v, int lo, int hi) => v < lo ? lo : (v > hi ? hi : v);

ReadinessBand _bandFor(int score) {
  if (score >= 70) return ReadinessBand.high;
  if (score >= 40) return ReadinessBand.moderate;
  return ReadinessBand.low;
}

ReadinessContribution? _scoreTsb(double? tsb) {
  if (tsb == null) return null;
  int delta;
  String note;
  if (tsb < -20) {
    delta = -20;
    note = 'Heavy fatigue — recent training stress is well above your fitness';
  } else if (tsb < -10) {
    delta = -12;
    note = 'Fatigued from recent training';
  } else if (tsb < -5) {
    delta = -6;
    note = 'Slight fatigue';
  } else if (tsb <= 5) {
    delta = 0;
    note = 'Form is neutral';
  } else if (tsb <= 15) {
    delta = 8;
    note = 'Fresh and well-recovered';
  } else if (tsb <= 25) {
    delta = 5;
    note = 'Very fresh — borderline tapered';
  } else {
    delta = -3;
    note = 'Over-tapered — edge may be blunted';
  }
  return ReadinessContribution(
      kind: ReadinessContributorKind.form, delta: delta, note: note);
}

ReadinessContribution? _scoreSleep(double? hours) {
  if (hours == null) return null;
  int delta;
  String note;
  if (hours < 5) {
    delta = -25;
    note = 'Very little sleep — recovery is compromised';
  } else if (hours < 6.5) {
    delta = -12;
    note = 'Short on sleep';
  } else if (hours < 7.5) {
    delta = -3;
    note = 'A little under target sleep';
  } else if (hours <= 9) {
    delta = 5;
    note = 'Well rested';
  } else {
    delta = 0;
    note = 'Extended sleep — recovery should be solid';
  }
  return ReadinessContribution(
      kind: ReadinessContributorKind.sleep, delta: delta, note: note);
}

ReadinessContribution? _scoreRestingHr(int? resting, int? baseline) {
  if (resting == null || baseline == null) return null;
  final diff = resting - baseline;
  int delta;
  String note;
  if (diff > 10) {
    delta = -18;
    note = 'Resting HR $diff bpm above baseline — strong sign of illness or under-recovery';
  } else if (diff > 5) {
    delta = -10;
    note = 'Resting HR $diff bpm above baseline';
  } else if (diff > 2) {
    delta = -4;
    note = 'Resting HR slightly elevated';
  } else if (diff >= -2) {
    delta = 0;
    note = 'Resting HR at baseline';
  } else {
    delta = 3;
    note = 'Resting HR below baseline — well recovered';
  }
  return ReadinessContribution(
      kind: ReadinessContributorKind.restingHr, delta: delta, note: note);
}

String _dominantAdvice(
    List<ReadinessContribution> contributors, ReadinessBand band) {
  if (contributors.isEmpty) {
    return band == ReadinessBand.high
        ? 'Looks like a good day to push the pace.'
        : 'Connect sleep + HR data for a real readiness picture.';
  }
  // Ties in |delta| are common — a TSB of -12 and six hours of sleep both
  // score -12, and the two carry DIFFERENT advice. Dart's List.sort is
  // unstable where JS's is stable, so without the index fallback the
  // dominant contributor here would be whichever the sort happened to land
  // on rather than the first one added, and the two platforms would show
  // different sentences for the same inputs.
  final indexed = <(int, ReadinessContribution)>[
    for (var i = 0; i < contributors.length; i++) (i, contributors[i]),
  ]..sort((a, b) {
      final c = b.$2.delta.abs().compareTo(a.$2.delta.abs());
      return c != 0 ? c : a.$1.compareTo(b.$1);
    });
  final dom = indexed.first.$2;
  final tail = switch (band) {
    ReadinessBand.low => 'Consider an easy day or a rest day.',
    ReadinessBand.moderate => 'A steady or moderate effort fits today.',
    ReadinessBand.high => 'You’re primed for a harder effort if planned.',
  };
  return '${dom.note}. $tail';
}

Readiness computeReadiness(ReadinessInputs inputs) {
  final contributors = <ReadinessContribution>[];
  final tsb = _scoreTsb(inputs.tsb);
  if (tsb != null) contributors.add(tsb);
  final sleep = _scoreSleep(inputs.sleepHours);
  if (sleep != null) contributors.add(sleep);
  final hr = _scoreRestingHr(inputs.restingHrBpm, inputs.baselineRestingHrBpm);
  if (hr != null) contributors.add(hr);

  final sum = contributors.fold<int>(0, (s, c) => s + c.delta);
  final score = _clamp(_baselineScore + sum, 0, 100);
  final band = _bandFor(score);
  return Readiness(
    score: score,
    band: band,
    advice: _dominantAdvice(contributors, band),
    contributors: contributors,
  );
}
